WL-3818/9: JS-Escaping and non-existent sites

nicholaswilson100 opened the pull request.

nicholaswilson100 committed:
WL-3819: JS-Escape the site contact name and site maintainer names

buckett commented:
How about doing this in bootstrap.jsp as shouldn't pretty much everything be escaped? For example i18n strings should be be escaped as well.

nicholaswilson100 commented:
yup okay

nicholaswilson100 commented:
I don't really want to be putting Java in jsp though and we don't seem to 'do' tags...

nicholaswilson100 commented:
We do some

buckett commented:
All I mean is that scattering escaping all over the code leads to things
slipping through. In this case everything that ends up in the JSP should be
escaped. Could do it in the .java file before putting anything in the
request.

On 12 February 2015 at 17:08, Nick Wilson notifications@github.com wrote:

> We do some
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ox-it/wl-feedback/pull/52#issuecomment-74109927.

## 

  Matthew Buckett, VLE Developer, IT Services, University of Oxford

nicholaswilson100 commented:
oh I see yup agreed

nicholaswilson100 commented:
Sorry I've pushed this 'WL-3818' Jira onto a branch here called 'WL-3819' which is another but related Jira.

nicholaswilson100 committed:
WL-3818: Escape everything else passed to handlebars

(Except "sakai.html.head" which broke the page when escaped and booleans and ints)

nicholaswilson100 commented:
If it's okay, I'm going to put Jira WL-3819 (non-existent sites) on this pull request too as it's dealing with the same files.

nicholaswilson100 committed:
WL-3819: EntityProviders do not like forward slashes in their post urls

If someone typed in a non-existent site like:

https://staging-werp.bsp.ox.ac.uk/portal/hierarchy/central/oucs/test/ada

and went to the Contact Us tool, then we create a form with a post action url of:

/direct/feedback/central/oucs/test/ada/reportcontent

which the EntityProvider backing the form does not like because it parses the url by forward slashes - which you can see from:

https://staging-werp.bsp.ox.ac.uk/direct/feedback/describe

So I'm replacing the forward slashes, if there are any, before and after the submission (which is ugly).

Using any of the methods of StringEscapeUtils or escaping the forward slashes does not work.

nicholaswilson100 commented:
(WL-3819 is) linked to https://github.com/ox-it/wl-tetraelf-hierarchy/pull/20

nicholaswilson100 committed:
WL-3819: Escape the value (==display name) not the key

buckett commented on src/java/org/sakaiproject/feedback/util/SakaiProxy.java:
```diff
@@ -4,6 +4,7 @@
 import java.text.MessageFormat;
 
 import org.apache.commons.fileupload.FileItem;
+import org.apache.commons.lang.StringEscapeUtils;
```
Unused.

buckett commented on src/java/org/sakaiproject/feedback/tool/entityproviders/FeedbackEntityProvider.java:
```diff
@@ -114,7 +115,7 @@ private String handleReport(final EntityView view, final Map<String, Object> par
         }
 
-        final String siteId = view.getPathSegment(1);
+        final String siteId = view.getPathSegment(1).replaceAll(FeedbackTool.FORWARD_SLASH, "/");
```
This sort of hack should have a note about why it's needed.

buckett commented on src/java/org/sakaiproject/feedback/tool/FeedbackTool.java:
```diff
@@ -140,12 +143,23 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
+		request.setAttribute(key, StringEscapeUtils.escapeJavaScript(value));
+	}
+
+	private void setMapAttribute(HttpServletRequest request, String key, Map map){
```
It's probably worth keeping the generics on the Map, so you can never get a unexpected map that contains nested maps.
